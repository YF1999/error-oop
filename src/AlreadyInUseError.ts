import { AlreadyInUseErrorMessageProps, AlreadyInUseErrorProps, ErrorOptions } from './CommonTypes';
import { _Error } from './NativeErrorsPro';

export class _AlreadyInUseError extends _Error {
    protected _entityName: string;

    public constructor(
        props: AlreadyInUseErrorProps,
        options: ErrorOptions<AlreadyInUseErrorMessageProps>,
    ) {
        const { entityName, inUse } = props;
        const { generateMessage: gm, ...others } = options;

        super(props, {
            generateMessage: gm && ((_props) => gm({ ..._props, entityName, inUse })),
            ...others,
        });

        this._entityName = entityName;
        this._setNonEnumerable('_entityName');
    }

    public get entityName() {
        return this._entityName;
    }
}

/**
 * Applicable when a resource is already in use, for example unique key constraints like a username.
 */
export class AlreadyInUseError extends _AlreadyInUseError {
    /**
     * @param entityName The entity that owns the protected resource.
     */
    public constructor(entityName: string);
    /**
     * @param entityName The entity that owns the protected resource.
     * @param arg1 A field or attribute that is already in use.
     */
    public constructor(entityName: string, arg1: string);
    /**
     * @param entityName The entity that owns the protected resource.
     * @param arg1 A field or attribute that is already in use.
     * @param arg2 A field or attribute that is already in use.
     */
    public constructor(entityName: string, arg1: string, arg2: string);
    /**
     * @param entityName The entity that owns the protected resource.
     * @param arg1 A field or attribute that is already in use.
     * @param arg2 A field or attribute that is already in use.
     * @param arg3 A field or attribute that is already in use.
     */
    public constructor(entityName: string, arg1: string, arg2: string, arg3: string);
    /**
     * @param entityName The entity that owns the protected resource.
     * @param arg1 A field or attribute that is already in use.
     * @param arg2 A field or attribute that is already in use.
     * @param arg3 A field or attribute that is already in use.
     * @param args Fields or attributes that are already in use.
     */
    public constructor(
        entityName: string,
        arg1: string,
        arg2: string,
        arg3: string,
        ...args: string[]
    );

    public constructor(
        entityName: string,
        arg1?: string,
        arg2?: string,
        arg3?: string,
        ...args: string[]
    ) {
        function generateMessage(props: AlreadyInUseErrorMessageProps) {
            return props.inUse.length === 0
                ? `The specified '${props.entityName}' value is already in use.`
                : `The specified '${props.entityName}' value is already in use:` +
                      `${props.inUse.join(', ')}`;
        }

        super(
            {
                message: '',
                entityName,
                inUse: [arg1, arg2, arg3, ...args].filter((e) => e !== undefined) as string[],
            },
            { generateMessage },
        );
    }
}
