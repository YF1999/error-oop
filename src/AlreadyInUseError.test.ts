import { AlreadyInUseError } from './AlreadyInUseError';

const entityName = 'entity';

testError(
    'AlreadyInUseError',
    () => {
        const err1 = new AlreadyInUseError(entityName);
        const err2 = new AlreadyInUseError(entityName, 'name');
        const err3 = new AlreadyInUseError(entityName, 'name', 'key');

        return [err1, err2, err3] as const;
    },
    (getErrors) => {
        test('property: message', () => {
            getErrors().map((err) => expectSnapshot(err.message));
        });

        test('property: inner error', () => {
            getErrors().map((err) => expect(err.innerError).toBeNull());
        });

        test('property: entity name', () => {
            getErrors().map((err) => expect(err.entityName).toBe(entityName));
        });
    },
);
