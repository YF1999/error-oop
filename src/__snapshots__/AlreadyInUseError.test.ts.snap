// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`AlreadyInUseError\` own keys 1`] = `
Array [
  "stack",
  "message",
  "_entityName",
  "_inUse",
]
`;

exports[`Test \`AlreadyInUseError\` own keys 2`] = `
Array [
  "stack",
  "message",
  "_entityName",
  "_inUse",
]
`;

exports[`Test \`AlreadyInUseError\` own keys 3`] = `
Array [
  "stack",
  "message",
  "_entityName",
  "_inUse",
]
`;

exports[`Test \`AlreadyInUseError\` property: message 1`] = `"The specified 'entity' value is already in use."`;

exports[`Test \`AlreadyInUseError\` property: message 2`] = `"The specified 'entity' value is already in use: name"`;

exports[`Test \`AlreadyInUseError\` property: message 3`] = `"The specified 'entity' value is already in use: name, key"`;
